presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java

```diff
@@ -325,133 +418,291 @@ private void calculateVariableValuesForLongConstraints(BoundVariables.Builder va
+        // If no item was seen ever, it provides UNCHANGED_SATISFIED.
+
+        private SolverReturnStatus current = SolverReturnStatus.UNCHANGED_SATISFIED;
+
+        public void add(SolverReturnStatus newStatus)
+        {
+            switch (newStatus) {
+                case UNCHANGED_SATISFIED:
+                    break;
+                case UNCHANGED_NOT_SATISFIED:
+                    if (current == SolverReturnStatus.UNCHANGED_SATISFIED) {
+                        current = SolverReturnStatus.UNCHANGED_NOT_SATISFIED;
+                    }
+                    break;
+                case CHANGED:
+                    if (current == SolverReturnStatus.UNCHANGED_SATISFIED || current == SolverReturnStatus.UNCHANGED_NOT_SATISFIED) {
+                        current = SolverReturnStatus.CHANGED;
+                    }
+                    break;
+                case UNSOLVABLE:
+                    current = SolverReturnStatus.UNSOLVABLE;
+                    break;
             }
-            return builder.build();
         }
-        return argumentSignatures;
+
+        public SolverReturnStatus getCurrent()
+        {
+            return current;
+        }
     }
 
-    public static List<TypeSignature> applyBoundVariables(List<TypeSignature> typeSignatures, BoundVariables boundVariables)
+    private class TypeParameterSolver
+            implements TypeConstraintSolver
     {
-        ImmutableList.Builder<TypeSignature> builder = ImmutableList.builder();
-        for (TypeSignature typeSignature : typeSignatures) {
-            builder.add(applyBoundVariables(typeSignature, boundVariables));
+        private final String typeParameter;
+        private final Type actualType;
+        private final boolean comparableRequired;
+        private final boolean orderableRequired;
+        private final Optional<String> requiredBaseName;
+
+        public TypeParameterSolver(String typeParameter, Type actualType, boolean comparableRequired, boolean orderableRequired, Optional<String> requiredBaseName)
+        {
+            this.typeParameter = typeParameter;
+            this.actualType = actualType;
+            this.comparableRequired = comparableRequired;
+            this.orderableRequired = orderableRequired;
+            this.requiredBaseName = requiredBaseName;
+        }
+
+        @Override
+        public SolverReturnStatus update(BoundVariables.Builder bindings)
+        {
+            if (!bindings.containsTypeVariable(typeParameter)) {
+                if (!satisfiesConstraints(actualType)) {
+                    return SolverReturnStatus.UNSOLVABLE;
+                }
+                bindings.setTypeVariable(typeParameter, actualType);
+                return SolverReturnStatus.CHANGED;
+            }
+            Type originalType = bindings.getTypeVariable(typeParameter);
+            Optional<Type> commonSuperType = typeManager.getCommonSuperType(originalType, actualType);
```

Comment:
As I pointed yesterday to @cberner and @martint, it might be needed for `getCommonSuperType` to behave differently in SQL context and lambda context. In SQL context character and possibly numeric types probably shouldn't have supertype (e. g. `char(5)` is not a supertype of `char(3)`, `double` is not a supertype of `decimal`). The difference in behavior is driven by SQL 99 standard compatibility.